Hash method

SkandanC opened the pull request:
Added __hash__() method in models.py.

SkandanC committed:
Added __hash__() method to models.py

SkandanC committed:
Add __hash__() function

SkandanC committed:
Create modelhashtest.py

SkandanC committed:
Added hash method to models.py

SkandanC committed:
Deleted test files

AustP reviewed (changes requested):
Here are some tests that you should test against:

```python
# wg1 and wg2 hash should be different
wg1 = Waveguide(length=150e-6)
wg2 = Waveguide(length=50e-6)

# wg1 and wg2 hash should be the same
wg1 = Waveguide(length=150e-6)
wg2 = Waveguide(length=150e-6)

# wg and gc hash should be different
wg1 = Waveguide(length=150e-6)
gc = GratingCoupler()
```

The model hash method doesn't need to worry about connections. A way to think about this is: what data needs to be encoded in order to determine if two distinct instances of a model represent the same component? In the case of waveguide, that will require the name/type of the component, and the length of it.

You may have to implement the __hash__ method in more places than just Model to ensure that all components work.

SkandanC committed:
Add __hash__() methods for siepic models

SkandanC committed:
Add __hash__() methods to sipann models

SkandanC committed:
Add _hash__() method to models.py

SkandanC committed:
Add hash tests

SkandanC committed:
Update __hash__() docstring

SkandanC committed:
Add base tests

SkandanC committed:
Update siepichashtest.py

SkandanC committed:
Update hash test files

Delete one of the hash test files and put tests into a class method under one file.

Change print statements to assert statements.

SkandanC committed:
Linting

SkandanC committed:
Linting

SkandanC committed:
Remove unused import

SkandanC committed:
Add SiPANN to tox deps

SiPANN is needed in the `py` virtual environment.

SkandanC committed:
Update test_hash.py

SkandanC committed:
Update assert statements

SkandanC committed:
Add SiPANN and numpy version to deps

SiPANN needed for SiPANN hash tests.

numpy version of at least 1.22 needed to avoid nlopt import issues (See here: https://githubhot.com/repo/stevengj/nlopt/issues/447)

SkandanC committed:
Update test_hash.py

SkandanC committed:
Update tox.ini

SkandanC committed:
Add numpy version and remove py37

Python 3,8 is required to install numpy>=1.22.0, which is needed yo avoid nlopt import errors while attempting to import sipann.

SkandanC committed:
Remove python version 3.7

Python 3.8+ is needed to install numpy>=1.22.0

SkandanC committed:
Remove numpy version requirement

SkandanC committed:
Update build-and-test.yml

SkandanC committed:
Temporarily disable SiPANN hash tests

SkandanC committed:
Uncomment SiPANN hash tests
